[JSC] A bit more optimizations for `n % power-of-two` in FTL

Constellation opened the pull request:
#### 3ec66a7fd7e54ddc27c98b3c51995f5ec9a8749a
<pre>
[JSC] A bit more optimizations for `n % power-of-two` in FTL
<a href="https://bugs.webkit.org/show_bug.cgi?id=302763">https://bugs.webkit.org/show_bug.cgi?id=302763</a>
<a href="https://rdar.apple.com/165021624">rdar://165021624</a>

Reviewed by Keith Miller.

We should do a bit more optimizations for `n % power-of-two`.

1. Int32 / Int32-power-of-two (larger than 1) never generates
   non-Int32 value except for negative 0 `-1 % 1`
2. Use DFG&apos;s ArithMod optimization to FTL.

Test: JSTests/stress/arith-modulo-power-of-two-int32.js

* JSTests/stress/arith-modulo-power-of-two-int32.js: Added.
(testMod2):
(testMod4):
(testMod8):
(testMod16):
(testMod32):
(testMod64):
(testMod128):
(testMod256):
(testMod512):
(testMod1024):
(testMod2048):
(testMod4096):
(testMod65536):
(testMod1048576):
(chainedMod):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileArithMod):

Canonical link: <a href="https://commits.webkit.org/303241@main">https://commits.webkit.org/303241@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a66b312936e8f204c370d7c803ed8bd2ccdaa704

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131801 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4293 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42807 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139312 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/83677 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b69f51a8-546b-495f-8633-7d5f0ef0e448) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4224 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4055 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/100739 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/83677 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134747 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3030 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118046 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81527 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/37885545-0061-4e71-94d1-c86475c5beea) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/2917 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/761 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82532 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/123864 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111656 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36171 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/141956 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/130308 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/3962 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36750 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109112 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4043 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/3473 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109276 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3010 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114327 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57172 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20490 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4016 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/32729 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/163274 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/3848 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67463 "Built successfully") | [⏳ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Tests-EWS "Waiting to run tests") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/4108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/3976 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -3240,6 +3240,53 @@ class LowerDFGToB3 {
+                    //
+                    // If numerator < 0:  result = divisor - 1
+                    // If numerator >= 0: result = 0
+                    unsigned logarithm = WTF::fastLog2(static_cast<uint32_t>(divisor));
```
Nit: I think this could just be a `ctz` (ditto) for the SpeculativeJIT code

Constellation replied:
Done.

kmiller68 reviewed (approved):
r=me with nit

Constellation committed:
[JSC] A bit more optimizations for `n % power-of-two` in FTL
https://bugs.webkit.org/show_bug.cgi?id=302763
rdar://165021624

Reviewed by Keith Miller.

We should do a bit more optimizations for `n % power-of-two`.

1. Int32 / Int32-power-of-two (larger than 1) never generates
   non-Int32 value except for negative 0 `-1 % 1`
2. Use DFG's ArithMod optimization to FTL.

Test: JSTests/stress/arith-modulo-power-of-two-int32.js

* JSTests/stress/arith-modulo-power-of-two-int32.js: Added.
(testMod2):
(testMod4):
(testMod8):
(testMod16):
(testMod32):
(testMod64):
(testMod128):
(testMod256):
(testMod512):
(testMod1024):
(testMod2048):
(testMod4096):
(testMod65536):
(testMod1048576):
(chainedMod):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileArithMod):

Canonical link: https://commits.webkit.org/303241@main
